### ISSUE TITLE
arXiv 25 | AREAL: A Large-Scale Asynchronous Reinforcement Learning System for Language Reasoning

### ISSUE THREAD

**Issue by hiahei** [AI-Infra-Team/awesome-papers#121]
# **AREAL: A Large-Scale Asynchronous Reinforcement Learning System for Language Reasoning**
Authors: 
Wei Fu, Jiaxuan Gao, Xujie Shen, Chen Zhu, Zhiyu Mei, Chuyi He, Shusheng Xu, Guo Wei, Jun Mei, Jiashu Wang, Tongkai Yang, Binhang Yuan, Yi Wu
[Paper](https://arxiv.org/abs/2505.24298)

### 1 Motivation

Reinforcement learning (RL) has emerged as a new scaling paradigm for enhancing the capabilities of large language models (LLMs) by enabling thinking abilities. **Effective RL training often requires massive parallelization to derive a large batch of rollouts for sufficient exploration**, which is the key to obtaining optimal model performance. For example, popular RL algorithms, such as PPO and GRPO, often require **an effective training batch of thousands of outputs**. Moreover, an LRM can generate tens of thousands of thinking tokens for each input prompt, further posing an urgent need for an efficient training system to run RL training on a large scale.

However, developing an efficient large-scale RL system is challenging. An RL system needs to frequently switch between LLM generation and training, which can introduce significant system overhead without careful optimizations. Furthermore, classical large-scale RL algorithms like PPO or GRPO typically require on-policy training data, i.e., samples generated by the latest model, to ensure the best model performance, which poses additional system challenges. 

In recent synchronous design, the generation step must wait until the finish of the longest output within a batch. Due to the varying output lengths for LRMs, a synchronous RL system suffers from severe training inefficiency.

<img width="1166" height="266" alt="Image" src="https://github.com/user-attachments/assets/b9ca7204-e80c-479d-9179-f90e8cab5bc7" />

> - Inference devices are underutilized.

> - Scalability is poor in synchronous RL systems.

We develop AREAL, a fully Asynchronous RL training system for LRMs that completely decouples generation from training without hurting the final performance. 
- AREAL runs LLM generation in **a streaming manner**, where each rollout worker continuously generates new outputs without waiting, leading to high GPU utilization. Meanwhile, the trainer workers in AREAL run **parallel model updates** whenever a training batch is obtained from the rollout workers. Once the model is updated, we synchronize the model weights in each rollout worker. 
- AREAL conducts a data filtering process to **ensure the staleness of each training sample is well controlled**. 
- AREAL introduces **several system-level optimizations**, including interruptible rollout workers, dynamic batching for variable-length outputs, and parallel reward service, which further improve the overall training throughput.

### 2 Method

<img width="1138" height="401" alt="Image" src="https://github.com/user-attachments/assets/5fb35417-81fe-44e2-9210-0838e48c090b" />
The system comprises 4 core components:

> - Interruptible Rollout Worker
> - Reward Service
> - Trainer Workers
> - Rollout Controller

While the asynchronous system design offers significant acceleration through improved device utilization, it introduces several technical challenges that require algorithmic considerations.

- Data Staleness: Staleness-Aware Training
We introduce a hyperparameter η representing the maximum permitted staleness in each training batch for stalenessaware training.

<img width="226" height="49" alt="Image" src="https://github.com/user-attachments/assets/1ede98d2-5a24-4f7d-91c7-3cad21d2f0af" />

- Inconsistent Policy Versions: Decoupled PPO Objective

<img width="827" height="180" alt="Image" src="https://github.com/user-attachments/assets/2ee1842c-b7e9-4d1a-95b5-cfe489d79259" />
<img width="774" height="34" alt="Image" src="https://github.com/user-attachments/assets/e7c04480-b08d-4beb-884d-92895b4171d2" />
By employing a recent policy as the proximal policy, model updates happen within the trust region around the high-quality proximal policy $π_{prox}$, thus stabilizing training.

### 3 Evaluation

Our evaluation comprises three components:
 (1) comprehensive comparisons against state-of-theart open-source frameworks across model sizes,
 (2) strong-scaling analysis with varying compute resources, 
 (3) ablation studies validating our design choices.

<img width="853" height="498" alt="Image" src="https://github.com/user-attachments/assets/2beeade1-3bda-4e09-a98c-e36ecf90d64b" />

<img width="855" height="390" alt="Image" src="https://github.com/user-attachments/assets/751756c5-6356-410e-af64-2ed00b082888" />

<img width="852" height="383" alt="Image" src="https://github.com/user-attachments/assets/ff2d2fa3-af52-4251-9e73-077ea736c3ec" />

<img width="852" height="279" alt="Image" src="https://github.com/user-attachments/assets/3c3eca28-581b-44a0-992c-24e7d50b68ac" />
